### PR TITLE
[battle, damage] refactor damage formula to match gen5+

### PR DESCRIPTION
**Pull request by SBird1337**
Refactors the damage calculation and fixes any (to my knowledge) inconsistancies to the calculation of later games.

## Description

Key notes are:

* Gamefreak often times rounds down on n.5 in their fixed point mathematics.
* fixed point arithmetic (multiplication) is not associative, this changes the order of operations to match the original games.
* A lot of the damage calculation function was quite messy, some aspects were factored in at the wrong place.
* The main damage calculation should now be clearer to read.

Also fixes the contrary test against intimidate, which was using a wrong multiplier.

Adds tests for the basic damage calculation using an example from https://bulbapedia.bulbagarden.net/wiki/Damage#Example
Changes the test in `weather_snow.c` against solar blade from `KNOWN_FAILING` to `PASS`.

~~At the current state 2 old tests are failing with these changes:~~
* ~~`Dry Skin increases damage taken from Fire-type moves by 25%`~~
* ~~`Swarm boosts Bug-type moves in a pinch`~~

Refactored the 2 problematic tests to use exact values.

this is due to numerical issues in the damage formula that also exist in the original games. I verfied both on showdown and accepted that as ground truth.

## **Discord contact info**
karathan

### CODE REVIEW

**Comment by SBird1337** [comment]
Technically I think this is ready for review, but I'll leave it as draft until we figure out what to do with the failing tests.

**Comment by SBird1337** [comment]
> Technically I think this is ready for review, but I'll leave it as draft until we figure out what to do with the failing tests.

I used exact values I obtained from a Smogon Calc for the 2 problematic tests. Ready for review :)

File: src/battle_util.c [hunk 1]
```diff
@@ -9393,270 +9375,409 @@ static u32 CalcDefenseStat(u16 move, u8 battlerAtk, u8 battlerDef, u8 moveType,
...
-        finalModifier = uq4_12_multiply(finalModifier, UQ_4_12(2.0));
-    if (gBattleMoves[move].damagesUnderwater            && gStatuses3[battlerDef] & STATUS3_UNDERWATER)
-        finalModifier = uq4_12_multiply(finalModifier, UQ_4_12(2.0));
-    if (gBattleMoves[move].damagesAirborneDoubleDamage  && gStatuses3[battlerDef] & STATUS3_ON_AIR)
-        finalModifier = uq4_12_multiply(finalModifier, UQ_4_12(2.0));
-
-    dmg = ApplyModifier(finalModifier, dmg);
-    if (dmg == 0)
-        dmg = 1;
+#define DAMAGE_ACCUMULATE_MULTIPLIER(modifier) do {                     \
```
Comment (AsparagusEduardo) [inline src/battle_util.c:9685]: Maybe `APPLY_DMG_MULTIPLIER` would be less verbose?

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: src/battle_util.c (hunk 1 above)
Comment (SBird1337) [reply src/battle_util.c:9685]: I'm not stuck to thes names by any means :D I tired to convey the following:

* `DAMAGE_APPLY_MODIFIER` will multiply an integer by a FP number and round (with gf rounding)
* `DAMAGE_ACCUMULATE_MULTIPLIER` will multiply a FP number by a FP number and round (with gf rounding)

the `apply` at least in my head signals this, while `accumulate` say we still have a FP number in the end.

Now that I think about it that is probably not clear to everyone, I'm open for suggestions :)

**Comment by SBird1337** [comment]
Please let me rebase before merging

File: src/battle_util.c (hunk 1 above)
Comment (CallmeEchoo) [reply src/battle_util.c:9685]: we use `MulModifier` for type effectiveness, so `DAMAGE_MUL_MULTIPLIER` might be more intuitive

Comment (CallmeEchoo) [reply src/battle_util.c:9685]: ah given that the variable is called `finalModifier` (which I dont object to) using `_MODIFIER` for the macros makes sense. 👍

Comment (SBird1337) [reply src/battle_util.c:9685]: I chose `DAMAGE_APPLY_MODIFIER` and `DAMAGE_MULTIPLY_MODIFIER` respectively :)

**Comment by SBird1337** [comment]
addressed Edus review and rebased.

File: test/ability_dry_skin.c [hunk 2]
```diff
@@ -44,7 +44,9 @@ SINGLE_BATTLE_TEST("Dry Skin increases damage taken from Fire-type moves by 25%"
         MESSAGE("Wobbuffet used Ember!");
         HP_BAR(opponent, captureDamage: &results[i].damage);
     } FINALLY {
-        EXPECT_MUL_EQ(results[0].damage, Q_4_12(1.25), results[1].damage);
+        // Values obtained from https://calc.pokemonshowdown.com
```
[the remark this answers is not in the document]
Comment (AsparagusEduardo) [reply test/ability_dry_skin.c]: You didn't include the assumptions that I requested.

[the remark this answers is not in the document]
Comment (SBird1337) [reply test/ability_dry_skin.c]: Hmm do you think we should rather assume the base stats or just hardcode the resulting A/D like in the general damage calculation.

[the remark this answers is not in the document]
Comment (AsparagusEduardo) [reply test/ability_dry_skin.c]: Yeah, forgot we could do this xD

**Review by AsparagusEduardo — APPROVED** [review]

**Review by mrgriffin — APPROVED** [review]

**Comment by mrgriffin** [comment]
@SBird1337 would you like to rebase, should I squash merge, or should I do a regular merge commit? :)

**Pull request merged**